### PR TITLE
Added check for isolated routable segments

### LINKED ISSUE #34: New check: No public connection for public segment
Public segment being defined as freeway, major highway, minor highway, primary street, street, ramp, dirt road.

This would highlight, for example, a public segment that only connected to a parking lot at either end, or a private road at either end. This would help to highlight errant public segments amidst large parking lots or private road networks.

### PR DESCRIPTION
**Pull request by davidakachaos**
As requested in #34 this added a check if a routable (and so public) segment
is isolated from other routable segments. If so, report that.
Check has been added to the I18n too in a seperate PR

Closes #34

### CODE REVIEW

**Comment by davidakachaos** [comment]
Hold off on merging this, I think there is something wrong. When I use the scanner it wrongly reports roads as being this, while when I go an look the report isn't there (and it shouldn't be there)

**Comment by berestovskyy** [comment]
Please give me a moment, I'll review the changes ASAP.

**Comment by davidakachaos** [comment]
Is there anything I can do to improve this PR?

File: src/validate.js [hunk 1]
```diff
@@ -502,6 +502,8 @@ function F_VALIDATE(disabledHL) {
 
 		this.$address = new _WV.SimpleADDRESS(attrs.primaryStreetID);
 
+		this.$isRoutable = this.$rawSegment.isRoutable();
```
Comment (berestovskyy) [inline src/validate.js]: $isRoutable should be read only, i.e. set property "writable: false" as we do for other fields.

File: src/validate.js [hunk 2]
```diff
@@ -2425,6 +2427,34 @@ function F_VALIDATE(disabledHL) {
...
+							break;
+						}
+					}
+				}
+				if (!foundPublicConnection
+					&& isLimitOk(202)
+					&& address.isOkFor(202))
+					segment.report(202);
+			}
+
```
Comment (berestovskyy) [inline src/validate.js]: Otherwise looks fine to me

File: src/validate.js [hunk 3]
```diff
@@ -2429,30 +2429,31 @@ function F_VALIDATE(disabledHL) {
 
 			// check Public connection
 			if (slowChecks
-				&& segment.$isRoutable){
+				&& segment.$isRoutable
+				&& nodeA.$otherSegments.length > 0
```
Comment (berestovskyy) [inline src/validate.js]: `$otherSegments` has a getter -- please see `getOtherSegments()`. Using `$otherSegments.length` will trigger the getter and hence will call segment constructors for all the segments connected to nodeA.

We better use `$otherSegmentsLen` to postpone (or maybe avoid) those calls.

File: src/validate.js [hunk 4]
```diff
@@ -2429,30 +2429,31 @@ function F_VALIDATE(disabledHL) {
...
+				for (var i = 0; i < nodeA.$otherSegments.length; i++) {
 					var otherSegment = nodeA.$otherSegments[i];
-					if (otherSegment.$isRoutable){
+					if (otherSegment.$rawSegment.isRoutable()){
 						foundPublicConnection = true;
 						break;
 					}
 				}
-				// if not already found, check node B
-				if(!foundPublicConnection){
```
Comment (berestovskyy) [inline src/validate.js]: This `if()` looks fine to me, i.e. we should avoid nodeB checks if there a public connection found on nodeA. Or did I miss something?

File: src/validate.js [hunk 5]
```diff
@@ -2429,30 +2429,31 @@ function F_VALIDATE(disabledHL) {
...
 			// check Public connection
 			if (slowChecks
-				&& segment.$isRoutable){
+				&& segment.$isRoutable
+				&& nodeA.$otherSegments.length > 0
+				&& nodeB.$otherSegments.length > 0){
 				// Check other segments to be a drivable public segment
 				var foundPublicConnection = false;
-				for (var i = 0; i < nodeA.$otherSegmentsLen; i++) {
+				for (var i = 0; i < nodeA.$otherSegments.length; i++) {
```
Comment (berestovskyy) [inline src/validate.js]: `$otherSegments.length` should have the same value as `$otherSegmentsLen`, but will be a bit slower due to the `$otherSegments` getter described above.

File: src/validate.js [hunk 6]
```diff
@@ -2429,30 +2429,31 @@ function F_VALIDATE(disabledHL) {
...
+						break;
 					}
 				}
 				if (!foundPublicConnection
 					&& isLimitOk(202)
-					&& address.isOkFor(202))
-					segment.report(202);
+					&& address.isOkFor(202)){
+						segment.report(202);
+					}
```
Comment (berestovskyy) [inline src/validate.js]: IMO the bug fix is the very first `if()`, i.e.:

    // check Public connection
    if (slowChecks
        && segment.$isRoutable
        && nodeA.$otherSegmentsLen > 0 // do not run the check if there are no connected segments


Have a look at `$isPartial`. It should be true if the node is not complete. I guess that is the issue you have encountered.

Please note that all the checks are grouped into chunks masked with `GROUP ...` comment. Basically, if you move the check into say a `GROUP isDrivable.!nodeApartial.!nodeBpartial` block, it should work OK without extra if conditions.

We might need to simplify/refactor this, but that is how it is at the moment...

**Comment by berestovskyy** [comment]
Hey David,
Sorry for the delay and thanks for the contribution!

There are few minor issues I think we can address. Let me know if you are up to this otherwise I can help.

**Comment by davidakachaos** [comment]
No problem. I would love to work on the comments you made, but today I'm
going into surgery. So I don't know if I'm able to help this week. But when
I'm back, I'll definitely look in to the changes for this.

I'm happy to help you with this project. You are kind and nice to work
with. Thank you!

Op ma 3 sep. 2018 00:55 schreef berestovskyy <notifications@github.com>:

**Comment by berestovskyy** [comment]
David,
No problem, it will wait for you.

The surgery, I hope it is nothing serious.

Get better ASAP, we need you ;)
Andriy

**Comment by davidakachaos** [comment]
Hey Andiry, I'm back! 😉 All wend well and I'm recovering well too.

I'll go over your remarks and see what I can do.

**Comment by davidakachaos** [comment]
@berestovskyy I've made changes according to your review. Also I noticed an update by URO+ regarding Waze.location no longer being present! I've added a fix for that in this PR. If you want, I can extract that to a seperate PR to fix?

**Comment by berestovskyy** [comment]
Merged manually

**Comment by davidakachaos** [comment]
Thank you! And I will try to be more tidy with my commits and squash them down a bit more before opening a PR 😄

**Pull request closed without merging**